rfd/0198-hardware-key-pin-caching.md

```diff
@@ -0,0 +1,189 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 198 - Hardware Key Pin Caching
+
+## Required Approvers
+
+* Engineering: @rosstimothy
+* Product: @xinding33 || @klizhentas
+
+## What
+
+Teleport offers the option to enforce the use of a hardware-backed private key
+to connect to Teleport servers and services, with an additional option to
+require the user's hardware key (PIV) pin for every operation
+(`hardware_key_pin`).
+
+This RFD proposes the implementation of a pin caching mechanism to improve UX
+when using the `hardware_key_pin` option.
+
+See [RFD 80](./0080-hardware-key-support.md) for more details on Hardware Key
+Support.
+
+## Why
+
+Although `hardware_key_pin` provides a great security option, it currently
+requires users to enter their pin once for every single action
+(e.g. `tsh ssh`, `tsh ls`). This is very disruptive when running several
+commands in short succession, especially when:
+
+* running several kubernetes or database commands through a Teleport local proxy.
+* using automated scripts which runs commands in bulk.
+
+## Details
+
+### UX
+
+Pin caching will significantly improve the UX of Hardware Key pin support by
+reducing the frequency of cumbersome pin prompts in the middle of active usage.
+Specifically, a user will be prompted at most once per configured timeout
+duration, rather than the the worst case scenario of being prompted every few
+seconds.
+
+Note that the Teleport client caching the pin will not prompt for pin
+immediately after the cache times out. Instead, it will prompt for pin the next
+time the user performs an action that would require pin.
+
+UX implications of the Teleport Key Agent will be covered in the Teleport Key
+Agent RFD.
+
+### Single Process Solution
```

Comment:
Will we implement both single and multi-process, or would we always use the key agent if it's available?

Reply:
Maybe add a diagram of how this plays out to the RFD?

> The Teleport Client will cache the PIN in memory

The Teleport Client (as in tsh) or the Key Agent? I'm confused.

> or remove the RFD altogether

I think it's worth capturing this interaction in an RFD - whether it's this one or RFD 199 I don't have strong opinions about.